File: solutions/lua/square-root/1/square-root.lua
local SquareRoot = {}

function SquareRoot.square_root(radicand)
    -- return radicand ^ 0.5
    function internal(xn)
      xn1 = 0.5 * (xn + radicand / xn)
      if (math.abs(xn - xn1) < 0.0000001) then
        return math.floor(xn1 + 0.5)
      else
        return internal(xn1)
      end
    end
    return internal(radicand)
end

return SquareRoot

File: solutions/elixir/square-root/1/lib/square_root.ex
defmodule SquareRoot do
  @doc """
  Calculate the integer square root of a positive integer
  """
  @spec calculate(radicand :: pos_integer) :: pos_integer
  def calculate(radicand) do
    calculate_internal(radicand, 1)
  end

  def calculate_internal(radicant, x) do
    if abs(x*x - radicant) == 0 do
      x
    else
      x_n = 0.5 * (x + radicant / x)
      calculate_internal(radicant, x_n)
    end
  end
end
